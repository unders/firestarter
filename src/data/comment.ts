import { CSS as css } from './css';

export class Comment {
   readonly body: string;

    constructor(body: string) {
       this.body = body;
    }
}

export class CommentError {
    body: string;
    err: string;

    constructor(body: string, err: string) {
        this.body = body;
        this.err = err;
    }
}

export class CommentFormWidget {
    placeholder: CommentFormPlaceholder;
    form: CommentForm;
    submit: CommentFormSubmit;
    data: CommentFormData;

    constructor(submitText: string, placeholderText: string) {
        this.placeholder = new CommentFormPlaceholder(placeholderText);
        this.form = new CommentForm();
        this.submit = new CommentFormSubmit(submitText);
        this.data = new CommentFormData();
    }
}

class CommentFormPlaceholder {
    readonly text: string = "Write a comment...";
    klass: string = css.show;
    constructor(text: string) {
        this.text = text;
    }
}

class CommentForm {
    klass: string = css.hide;
}

class CommentFormSubmit {
    readonly title: string = "Publish";
    disable: boolean = false;

    constructor(title: string) {
        this.title = title
    }
}

class CommentFormData {
    [key: string]: string;
    body: string = "";
}

