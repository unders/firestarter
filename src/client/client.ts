import { Error } from '../data/error';
import { Comment } from '../data/comment';
import { Env } from "../data/env";

export class Response {
    readonly json: string;
    readonly err: Error | null;

    constructor(json: string, err: Error | null) {
        this.json = json;
        this.err = err;
    }
}

export interface XHR {

    /**
     * @example
     *          const {json, err} = await xhr.post(path, json));
     *          if (err) {
     *              ...
     *          } else {
     *              ...
     *          }
     *
     */
    post(path: string, json: string): Promise<Response>;
}

export class Headers {
    [key: string]: string;
}

//
// https://xhr.spec.whatwg.org/
//
export class Client {
    readonly timeout: number = 5000; // time in milliseconds (default = 5 seconds)
    readonly headers: Headers;
    readonly host: string; // http://localhost:8080

    constructor(host:string, headers: Headers, timeout: number) {
        this.host = host;
        this.headers = headers;
        this.headers['Content-Type'] = 'application/json; charset=utf-8';
        if (timeout !== 0) {
            this.timeout = timeout;
        }
    }

    static make(env: Env, host: string, headers: Headers, timeout: number): XHR {
        if (!env.isProd()) {
            return new MockClient();
        }
        return new Client(host, headers, timeout);
    }

    async post(path: string, json: string): Promise<Response> {
        return new Promise<Response>((resolve) => {
            try {
                const xhr = new XMLHttpRequest();
                for (let key in this.headers) {
                    xhr.setRequestHeader(key, this.headers[key]);
                }
                xhr.timeout = this.timeout;
                xhr.open("POST", this.host + path);

                xhr.ontimeout = (e: ProgressEvent) => {
                    e.preventDefault();
                    const header = "Request Timeout";
                    const message = "Request took to long, please try again.";
                    const err = new Error(503, header, message);
                    const resp = new Response("{}", err);
                    resolve(resp);
                };

                xhr.onloadend = (e: ProgressEvent) => {
                    e.preventDefault();

                    if (xhr.status < 300) {
                        // 200 ... 299
                        const resp = new Response(xhr.responseText, null);
                        resolve(resp);
                    } else if (xhr.status > 299 && xhr.status < 500) {
                        // 400 .. 499
                        const err =  Error.fromJSON(xhr.responseText);
                        const resp = new Response("{}", err);
                        resolve(resp);
                    } else if (xhr.status === 503) {
                        const header = "Request Timeout";
                        const message = "Request took to long, please try again.";
                        const err = new Error(xhr.status, header, message);
                        const resp = new Response("{}", err);
                        resolve(resp);
                    } else {
                        const header = "A server problem";
                        const message = "It was a problem on our server, please try again.";
                        const err = new Error(xhr.status, header, message);
                        const resp = new Response("{}", err);
                        resolve(resp);
                    }
                };

                xhr.send(json);
            } catch (e) {
                // TODO: log the error
                const header = "The Request failed";
                const message = "We are investigating the problem, please try again.";
                const err = new Error(500, header, message);
                const resp = new Response("{}", err);
                resolve(resp);
            }
        })
    }
}

interface WithBody {
    body: string;
}

class MockClient {
    post(path: string, json: string): Promise<Response> {
        return new Promise<Response>((resolve) => {
            const data = JSON.parse(json) as WithBody;

            switch(data.body) {
                case "400": {
                    const header = "Your comment is not valid, you can do better";
                    const message = "400 is a number and that is not a valid comment.";
                    const err = new Error(400, header, message);
                    const resp = new Response("{}", err);
                    resolve(resp);
                    break;
                }
                default: {
                    // When it works
                    const comment = new Comment(data.body);
                    const resp = new Response(JSON.stringify(comment), null);
                    setTimeout(resolve, 1000, resp);
                    break;
                }
            }
        });
    }
}
