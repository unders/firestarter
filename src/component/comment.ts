import { Comment } from "../data/comment"
import { Poster} from "../client/client"
import { Dom as dom } from "../dom/dom"

export interface Props {
    root: string;
    listTitle: string;
    client: Poster;
    comments: Comment[];
}

export class Component {
    root: Element;
    form: Form;
    list: List;
    bind: (template: TemplateStringsArray, ...args : any[]) => void;

    constructor(props: Props) {
        const root = document.querySelector(props.root);

        if (root) {
            this.root = root;
            this.bind = dom.bind(this.root);
            this.form = new Form(props.client, new View(root));
            this.list = new List(props.listTitle, props.comments);

            this.render();
        }
    }

    render() {
        if (this.root) {
            this.bind`${[
                this.form.render()
                // this.list.render()
            ]}`;
        }
    }
}

class View {
    root: Element;
    data: Data = new Data();
    body: Body = new Body();
    submit: Submit = new Submit();

    constructor(root: Element) {
        this.root = root;
        this.resetBodySize = this.resetBodySize.bind(this);
    }

    focusBody() {
        const el = this.root.querySelector(".funcbox-textarea") as HTMLTextAreaElement;
        el.focus();
    }

    resetBodySize() {
        const el = this.root.querySelector(".funcbox-textarea") as HTMLTextAreaElement;
        el.style.height = "inherit";
        el.style.height = el.scrollHeight + "px";
    }

    isValid(): boolean {
        return this.body.isValid(this.data.body);
    }

    toJSON(): string {
        return JSON.stringify(this.data);
    }

    disableSubmit(): void {
        this.submit.disable = true;
    }

    enableSubmit(): void {
        this.submit.disable = false;
    }

    reset(): void {
        this.data.body = "";
        this.body.setOk();
        this.submit.disable = false;
        setTimeout(this.resetBodySize, 1);
    }
}
class Data {
    [key: string]: string;
    body: string = "";
}
class Body {
    static readonly ok: string = "";
    static readonly error: string = "error";
    errorKlass: string = "";
    errorMsg: string = "Please, write something.";
    placeholder: string = "Write a comment";

    isValid(value: string): boolean {
       if (value === "") {
           return this.setError();
       } else {
           return this.setOk();
       }
    }

    setError() {
        this.errorKlass = Body.error;
        return false;
    }

    setOk() {
        this.errorKlass = Body.ok;
        return true
    }
}
class Submit {
    readonly title: string = "Send";
    disable: boolean = false;
}


class Form {
    readonly client: Poster;
    readonly html: (template: TemplateStringsArray, ...args : any[]) => void;
    view: View;
    bodyHasGrow: boolean = false;

    constructor(client: Poster, view: View) {
        this.view = view;
        this.client = client;
        this.html = dom.wire(this);
        this.submit = this.submit.bind(this);
        this.saveInput = this.saveInput.bind(this);
        this.removeError = this.removeError.bind(this);
        this.grow = this.grow.bind(this);
        // this.shrink = this.shrink.bind(this);
    }

    saveInput(e: Event) {
        const input = e.target as HTMLInputElement;
        this.view.data[input.name] = input.value.trim();
    }

    removeError(e: Event) {
        if (this.view.body.errorKlass !== Body.ok) {
            this.view.body.errorKlass = Body.ok;
            this.render();
        }
    }

    grow(e: Event) {
        const el = e.target as HTMLTextAreaElement;

        if (!this.bodyHasGrow && el.scrollHeight > el.clientHeight) {
            this.bodyHasGrow = true;
            el.style.height = el.scrollHeight + "px";
        }

        if (this.bodyHasGrow) {
            el.style.height = "inherit";
            el.style.height = el.scrollHeight + "px";
        }
    }

    async submit(e: Event) {
        e.preventDefault();

        this.view.disableSubmit();
        this.render();

        if(!this.view.isValid()) {
            this.view.enableSubmit();
            this.render();
            this.view.focusBody();
            return;
        }
        const req = this.view.toJSON();
        this.view.reset();
        this.render();

        // animate a fast add at the top of list...

        const {json, err} = await this.client.post(req);
        if (err) {
            // animate and replace list item with error message
            if (err.code == 400) {

            }
            console.log(err.code, err.status, err.message, err.value);
        } else {
            // resolve optimistic update
            // update store.
            console.log("json: ", json);
        }
    }

    render() {
        const data = this.view.data;
        const body = this.view.body;
        const submit = this.view.submit;

        return this.html`
            <form   class="funcbox-comment-form"
                    onsubmit=${this.submit}
                    oninput=${this.saveInput}>
                <div class="${['funcbox-comment-group', body.errorKlass].join(' ')}">
                    <textarea   class="funcbox-textarea"
                                oninput="${this.grow}"
                                onkeydown="${this.removeError}"
                                tabindex="1"
                                name='body'
                                placeholder="${body.placeholder}"
                                value="${data.body}"></textarea>
                    <span class="funcbox-textarea-error">${body.errorMsg}</span>
                </div>
                <button class="funcbox-button"
                        tabindex="2"
                        disabled=${submit.disable}>${submit.title}</button>
            </form>`;
    }
}

class List {
    title: string;
    comments: Comment[];
    html: (template: TemplateStringsArray, ...args : any[]) => void;

    constructor(title: string, comments: Comment[]) {
        this.title = title;
        this.html = dom.wire(this);
    }

    click(e: Event) {
        console.log(e);
        console.log(e.target);
    }

    render() {
        return this.html`
            <h3>${this.title}</h3>
            <ul class="comments">${this.comments.map( (comment) => dom.wire(comment)`
                <li onclick="${this.click}">${comment.body}</li>`)}
            </ul>`;
    }
}
