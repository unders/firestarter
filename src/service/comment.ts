import { Router as router} from "../router"
import { XHR } from "../client/client"
import {IState, State} from "../data/state"
import {CSS as css} from "../data/css"
import {Comment, CommentListItem} from "../data/comment"

export class CommentService {
    private readonly xhr: XHR;
    private readonly comments: Comment[];
    private state: IState;

    constructor(xhr: XHR, comments: Comment[]) {
        this.xhr = xhr;
        this.comments = comments;
    }

    init(state: IState) {
        this.state = state;

        const listItems: CommentListItem[] = [];
        this.comments.forEach((comment) => {
            const listItem = new CommentListItem(comment, "", "", "");
            listItems.push(listItem);
        });

        this.state.getState().commentListWidget.comments = listItems;
    }

    async submitComment(comment: Comment, timeout: number) {
        const wait = this.minWait(timeout);
        const {json, err} = await this.xhr.post(router.commentPath, comment.toJSON());

        if (err) {
            const listItem = new CommentListItem(comment, css.errHighlight, err.header, err.message);

            await wait;
            this.state.setState((state: State): any => {
                state.commentListWidget.comments.unshift(listItem);
            });
            return;
        }

        const listItem = new CommentListItem(comment, css.highlight, "", "");

        await wait;
        this.state.setState((state: State): any => {
            state.commentListWidget.comments.unshift(listItem);
        });
    }

    private minWait(timeout: number): Promise<any> {
        return new Promise<any>((resolve) => {
            if (timeout === 0) {
                resolve();
            } else {
                setTimeout(resolve, timeout);
            }
        });
    }
}
