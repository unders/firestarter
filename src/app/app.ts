import { State } from "../data/state"
import { CommentComponent } from "../component/comment"
import { CommentService } from "../service/comment";

export class App {
    readonly state: State;
    readonly comment: CommentComponent;

    constructor(commentService: CommentService) {
        this.state = State.init();
        commentService.init(this);

        const comment = document.querySelector("#funcbox-comment");
        this.comment = new CommentComponent({
                root: comment,
                commentService: commentService,
                state: this });
    }

    setState(callback: (s: State) => {}): void {
        callback(this.state);
        this.comment.onStateChange();
        this.render();
    }

    getState(): State {
        return this.state
    }

    render() {
        this.comment.render();
    }
}
