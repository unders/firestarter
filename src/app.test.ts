import {Client, Headers} from "./client/client";
import {App} from "./app/app";
import {CommentService} from "./service/comment";
import {adjustSnap} from "./test/helper";
import {CommentComponent} from "./component/comment";

class Page {
    showForm() {

    }
}

describe("App", () => {
    document.body.innerHTML = '<div id="funcbox-comment"></div>';
    CommentComponent.timeout = () => { return 0; };
    const client = new Client("", new Headers(), 0);
    const commentService = new CommentService(client, []);
    const app = new App(commentService);

    test("#render()", () => {
        app.render();
        expect(adjustSnap(document.body.innerHTML)).toMatchSnapshot();
    });
});
