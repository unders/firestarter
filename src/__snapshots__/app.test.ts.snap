// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App #render() 1`] = `
"<div id=\\"funcbox-comment\\"><h3 class=\\"funcbox-comment-heading\\">Comments</h3><div class=\\"funcbox-placeholder \\" data-showform=\\"1\\">
                        <span class=\\"funcbox-placeholder-text\\">
                            Write a comment...
                        </span>
                    </div><form class=\\"funcbox-comment-form hide\\">
                        <textarea class=\\"funcbox-comment-textarea\\" name=\\"body\\" data-body=\\"1\\" placeholder=\\"Write a comment...\\"></textarea>
                        <div class=\\"funcbox-comment-footer\\">
                            <button class=\\"funcbox-comment-submit\\" data-cancel=\\"1\\">Cancel</button>
                            <button class=\\"funcbox-comment-submit publish\\" data-submit=\\"1\\">Publish</button>
                        </div>
                    </form><ul class=\\"funcbox-comment-list\\">
                <!--;-->
            </ul><!--;--></div>"
`;
